rfd/0057-automatic-user-provisioning.md

```diff
@@ -0,0 +1,123 @@
+kind: role
+version: v5
+metadata:
+  name: example
+spec:
+  options:
+    # Controls whether this role supports auto provisioning of users.
+    auto_create_user: true
+    # Controls whether this role allows adding auto provisioned users to sudoers.
+    add_to_sudoers: true
+  allow:
+    # New field listing Linux groups to assign a provisioned user to.
+    # Should support user and identity provider traits like other fields (e.g. "logins")
+    groups: [ubuntu, "{{internal.groups}}", "{{external.xxx}}"]
+```
+
+An individual `ssh_service` can be configured disable auto user
+creation with the below config:
+
+```yaml
+ssh_service:
+    # when disabled, takes precedence over the role setting
+    disable_auto_create_user: true
+```
+
+### User creation
+
+A PAM module can be used in order to avoid running a Teleport process
+as root, and is already recommended in the [Teleport docs](https://goteleport.com/docs/server-access/guides/ssh-pam/#create-local-unix-users-on-login).
+
+This script will be adapted to support multiple Linux distributions
+and their differing methods by which users are created, and can be
+distributed as part of released Teleport packages.
+
+#### User Groups
+
+When a user is created they will be added to the specified groups from
+the role. In addition the user will be added to a special
+`teleport-tmp` group which can be used to indicate that the user was
+created by teleport and that its safe for it to be deleted.
+
+#### Valid user/group names
+
+The set of valid names that are valid on Linux varies between distros
+and are generally more restrictive than the allowed usernames in
+Teleport. This will require that names containing invalid characters
+have those characters removed/replaced. Information on the valid
+characters between Linux distros is available [here](https://systemd.io/USER_NAMES/).
+The common core of valid characters is `^[a-z][a-z0-9-]{0,30}$`.
+
+#### Adding and removing users from sudoers
+
+A user can be added to `sudoers` by creating files for each user in
+`/etc/sudoers.d/`.
+
+Each file will contain the following contents, allowing the specified
+user to execute root commands without a password:
+```
+${USERNAME} ALL=(ALL) ALL
+```
+
+##### Other considerations
+
+`sudo` isn't always available and [doas](https://man.openbsd.org/doas.conf.5)
+can be used similarly to `sudo`.  Supporting systems using `doas`
+would required modifying `/etc/doas.conf` to include `permit` line:
+
+```
+permit nopass ${USERNAME} as root
+```
+
+On nodes with only `su` available a PAM module is available that
+allows for users in a particular group -- such as `sudoers` or
+`wheel`, may execute `su` commands.
+
+### User deletion
+
+After all of a users sessions are logged out the created user and any
+`sudoers` files that were created for that user will be deleted and
+the users account will also be deleted, this can also be managed by
+PAM.
+
+Users can not be deleted while they have running processes so each
```

Comment:
This is a good point, what happens if you launch a background process as an auto-created user and then exit the session?

In addition to trying to clean up the user upon the exit, can Teleport node keep track of existing sessions for users and have a background goroutine that would try to clean up all automatically created users without active sessions every once in a while?

Reply:
I dont think the user will be successfully deleted. A goroutine for cleanup of leftover users with no connections would make sense to kill all their processes and clean up.